Restored and updated the api toggle + added a download option for poster images

ltctceplrm opened the pull request:
This will allow users to choose which API to use per media type, speeding up the requests and also getting rid of missing API key errors by simply disabling that API if they're not interested in using it.

In this updated version the settings are now dynamically added rather than manually.

The API toggles are grouped per media type and look like this:
![image](https://github.com/user-attachments/assets/e88aed57-4312-4f73-93ee-9b1185dd59a9)

mProjectsCode commented on src/settings/Settings.ts:
```diff
@@ -310,82 +327,39 @@ export class MediaDbSettingTab extends PluginSettingTab {
+		const mediaTypeApiMap = new Map<string, string[]>();
+
+		// Populate the map with APIs for each media type
+		Object.entries(this.plugin.settings.apiToggle).forEach(([apiName, api]) => {
```
prefer a normal for loop over `.forEach(...)`

mProjectsCode commented on src/settings/Settings.ts:
```diff
@@ -21,24 +21,32 @@ export interface MediaDbPluginSettings {
-	// 		game: boolean;
-	// 	};
-	// };
+	apiToggle: {
```
true, if you don't wanna do that, it's fine. I can do it too, but no guarantee when I find the time for it.

ltctceplrm replied:
Well I tried doing it anyway, I spent more than two hours on it because I tried to have the new apiMediaTypes to auto populate with each mediatype per api but in the I couldn't figure that out so I just hardcoded
```
		const apiMediaTypes = {
			OMDbAPI: ['movie', 'series', 'game'],
			MALAPI: ['movie', 'series'],
			MALAPIManga: ['comicManga'],
			ComicVineAPI: ['comicManga'],
			SteamAPI: ['game'],
			MobyGamesAPI: ['game'],
			GiantBombAPI: ['game'],
		};
```

If I understand what you wanted correctly then this is still a good and easy way to add/modify mediatypes for an api.

"MALAPI Manga" gave me a lot of grief too before I figured out that it was the only API with a space in it's name and therefore wouldn't work. For now I just added a regex to remove the space but maybe it should be renamed to MALAPIManga?

ltctceplrm commented:
I reverted the earlier fix for "MALAPI Manga" with the space in it's name because it didn't work well, to ensure compatibility I think spaces shouldn't be allowed in API names. I've added a conversion from `dataSource: MALAPI Manga` to `dataSource: MALAPIManga` so it shouldn't break anything for existing users.

If you prefer me to try to make it work with the space instead I can try that again.

Otherwise it should be fully functional now

mProjectsCode commented on src/api/APIModel.ts:
```diff
@@ -19,13 +19,8 @@ export abstract class APIModel {
-		// 	return true;
-		// }
-		return this.types.contains(type);
+		const disabledMediaTypes = this.plugin.settings[`${this.apiName}_disabledMediaTypes` as keyof typeof this.plugin.settings] as MediaType[];
```
I am not the biggest fan of using an interpolated string to access the settings. This seems very error prone. Maybe an abstract method that every concrete API overrides is a better solution.

mProjectsCode replied:
Then the MALAPI name fix is not needed as well.

ltctceplrm replied:
Okay if I understood you correctly then I think I succeeded but it does require all APIs to have the override even if they're the only API for that media type, so I had to add a setting array for those too

ltctceplrm replied:
I also reverted the MALAPI Manga rename

mProjectsCode commented on src/settings/Settings.ts:
```diff
@@ -310,82 +289,55 @@ export class MediaDbSettingTab extends PluginSettingTab {
-		//	});
+		// Create a map to store APIs for each media type
+		const mediaTypeApiMap = new Map<string, string[]>();
+		const apiMediaTypes = {
```
Isn't this info already available on the API classes? Why not take it from there? Duplicated information is prone to errors when things change in the future.

ltctceplrm replied:
I kind of had a brain fart when I first tried to do it dynamically and gave up on that but I hadn't thought of getting the info from this.plugin.apiManager.apis instead. Good point

mProjectsCode reviewed (changes requested):
Some few more minor things.

mProjectsCode commented on src/settings/Settings.ts:
```diff
@@ -21,24 +22,17 @@ export interface MediaDbPluginSettings {
-	// 		game: boolean;
-	// 	};
-	// };
+	OMDbAPI_disabledMediaTypes: string[];
```
I would change all these to use `MediaType[]` instead of `string[]`. That saves you some casting later.

ltctceplrm replied:
Good point, thanks

mProjectsCode commented on src/settings/Settings.ts:
```diff
@@ -310,82 +298,49 @@ export class MediaDbSettingTab extends PluginSettingTab {
-		//		});
-		//	});
+		// Create a map to store APIs for each media type
+		const mediaTypeApiMap = new Map<string, string[]>();
```
Same thing here about using `MediaType` instead of `string`.

mProjectsCode commented on src/settings/Settings.ts:
```diff
@@ -310,82 +298,49 @@ export class MediaDbSettingTab extends PluginSettingTab {
+
+		// Populate the map with APIs for each media type dynamically
+		for (const api of this.plugin.apiManager.apis) {
+			for (const MediaType of api.types) {
```
Please use lowercase names for variables. Uppercase names are reserved for class names and types.

ltctceplrm replied:
If I use camelCase like mediaType it's fine though right?

ltctceplrm commented:
I'm adding a small bug fix for an unrelated issue: #188 where pressing cancel still creates the selected media entry, please let me know if you'd prefer it as a separate PR.

ltctceplrm commented:
The new download option would close #150, it is toggleable and defaults to false so if users prefer keeping the url they don't have to change anything, if users want to download then they can enter a path and turn on the button
![image](https://github.com/user-attachments/assets/74dc6028-109e-45b9-9fbb-42944a82346f)

Posters are saved as `"[[{{ folder path}}/{{ mediatype }}_{{ title }} {{ year }}.{{ ext }}]]"` which would translate to `"[[Resources/Posters/series_Day Break (2006–2007).jpg]]"`, I use the wikilinks this way because it's how obsidian links to files and the user can then easily modify or delete the poster if they so choose.

Users can also choose to embed the image in their note with this dataview:
`="!" + this.image` or `="![[" + meta(this.image).path + "|350]]"` if they want a specific size for the embedded image

![image](https://github.com/user-attachments/assets/e80b8005-5944-4621-885a-7aab07ac8706)

The poster resolution has been upgraded from a width of 300px to 600px and the steam posters are now vertical rather than horizontal to fit in with the rest of the posters.

Previous:
![header](https://github.com/user-attachments/assets/d6cb7d5e-cea2-400e-a7db-861e85b8363f)
New:
![image](https://github.com/user-attachments/assets/337686a5-a32f-479d-9dc2-3492ce652fa1)

mProjectsCode commented on src/main.ts:
```diff
@@ -306,6 +315,33 @@ export default class MediaDbPlugin extends Plugin {
+			if (mediaTypeModel.image && typeof mediaTypeModel.image === 'string' && mediaTypeModel.image.startsWith('http')) {
+				if (this.settings.imageDownload) {
+					try {
+						const imageurl = mediaTypeModel.image;
```
I would extract this into a separate method.

mProjectsCode commented:
Looks good, thanks :)
